### PR TITLE
Allow tangent on arcs/lines/splines without coincident endpoints.

### PR DESCRIPTION
**Pull request by ruevs**
Implements #937

Test cases:
[ArcsLinesSplinesTangentTests.zip](https://github.com/user-attachments/files/17648175/ArcsLinesSplinesTangentTests.zip)
https://github.com/solvespace/solvespace/pull/833#issuecomment-742082682
https://github.com/solvespace/solvespace/pull/833#issuecomment-744528586

Cross references:
https://github.com/solvespace/solvespace/pull/833
96958f466395148ee7f0a252222a2a27e7345c22
https://github.com/solvespace/solvespace/issues/937
https://github.com/solvespace/solvespace/pull/948
8c101d5f281ba670a3af2663302fe3dc7fb15283
https://github.com/solvespace/solvespace/issues/951
https://github.com/solvespace/solvespace/issues/1492

### CODE REVIEW

**Comment by phkahler** [comment]
Does this require you to select the end point of the arc? I would hope to automatically figure that out if it's constrained to the line. It's not too hard to search the constraints to figure it out?

**Comment by ruevs** [comment]
It does require one to select the end point(s) of the curve(s) at which the tangency constraint will be applied.

I tried to make the error messages as informative as possible - take a look or play with it.

If an arc/spline already has one point constrained on the line (not on it's end point) it does seem logical to automatically select that point, but it is not the only way - the other end of the curve can be constrained tangent. Auto selecting would constrain the generality of the solution, so I did not attempt to do it. Of cource if one explicitly selected a (the other) point we could override the automatic selection, but I think it becomes confusing.

File: src/constraint.cpp [hunk 1]
```diff
@@ -140,16 +141,32 @@ bool Constraint::ConstrainArcLineTangent(Constraint *c, Entity *line, Entity *ar
...
+    } else if(nullptr != arcendpoint) {
+        Vector p = arcendpoint->PointGetNum();
+        if(a1.Equals(p)) {
+            c->other = false;
+        } else if(a2.Equals(p)) {
+            c->other = true;
+        } else {
+            Error(_("The point you selected does not belong to the arc. "
+                    "The arc and line segment do not share an end point.\n\n"
+                    "Select the end point of the ark at which you want "
```
Comment (hofmannu) [inline src/constraint.cpp:153]: arc?

File: src/constraint.cpp [hunk 2]
```diff
@@ -140,16 +141,32 @@ bool Constraint::ConstrainArcLineTangent(Constraint *c, Entity *line, Entity *ar
...
+                    "it to be tangent to the line."));
+            return false;
+        }
     } else {
         Error(_("The tangent arc and line segment must share an "
                 "endpoint. Constrain them with Constrain -> "
-                "On Point before constraining tangent."));
+                "On Point before constraining tangent.\n\n"
+                "Alternatively select the end point of the ark "
```
Comment (hofmannu) [inline src/constraint.cpp:161]: arc

File: src/constraint.cpp (hunk 1 above)
Comment (ruevs) [reply src/constraint.cpp:153]: :-) Indeed, Noah is not involved :-)

File: src/constraint.cpp (hunk 2 above)
Comment (ruevs) [reply src/constraint.cpp]: Thanks for the review. I removed the arks and force pushed the branch.

**Comment by ruevs** [comment]
@phkahler so? 2b|!2b ?

**Comment by phkahler** [comment]
> @phkahler so? 2b|!2b ?

I'm just going to trust your judgement on this one. The code looks reasonable ;-)

**Comment by ruevs** [comment]
Merged then :-)

**Pull request merged**